libraries/AP_Compass/AP_Compass_IIS2MDC.cpp

```diff
@@ -0,0 +1,181 @@
+        enum Rotation rotation)
+{
+    if (!dev) {
+        return nullptr;
+    }
+
+    AP_Compass_IIS2MDC *sensor = NEW_NOTHROW AP_Compass_IIS2MDC(std::move(dev),force_external,rotation);
+
+    if (!sensor || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+AP_Compass_IIS2MDC::AP_Compass_IIS2MDC(AP_HAL::OwnPtr<AP_HAL::Device> dev,
+        bool force_external,
+        enum Rotation rotation)
+    : _dev(std::move(dev))
+    , _rotation(rotation)
+    , _force_external(force_external)
+{
+}
+
+bool AP_Compass_IIS2MDC::init()
+{
+    _dev->get_semaphore()->take_blocking();
+
+    _dev->set_retries(10);
+
+    if (!check_whoami()) {
+        goto fail;
+    }
+
+    if (!_dev->write_register(IIS2MDC_ADDR_CFG_REG_A, MD_CONTINUOUS | ODR_100 | COMP_TEMP_EN)) {
+        goto fail;
+    }
+
+    if (!_dev->write_register(IIS2MDC_ADDR_CFG_REG_B, OFF_CANC)) {
+        goto fail;
+    }
+
+    if (!_dev->write_register(IIS2MDC_ADDR_CFG_REG_C, BDU)) {
+        goto fail;
+    }
+
+    // lower retries for run
+    _dev->set_retries(3);
+
+    _dev->get_semaphore()->give();
+
+    // register compass instance
+    _dev->set_device_type(DEVTYPE_IIS2MDC);
+
+    if (!register_compass(_dev->get_bus_id(), _instance)) {
+        return false;
+    }
+
+    set_dev_id(_instance, _dev->get_bus_id());
+
+    printf("%s found on bus %u id %u address 0x%02x\n", name,
+           _dev->bus_num(), unsigned(_dev->get_bus_id()), _dev->get_bus_address());
+
+    set_rotation(_instance, _rotation);
+
+    if (_force_external) {
+        set_external(_instance, true);
+    }
+
+    // Enable 100HZ
+    _dev->register_periodic_callback(10000, FUNCTOR_BIND_MEMBER(&AP_Compass_IIS2MDC::timer, void));
+
+    return true;
+
+fail:
+    _dev->get_semaphore()->give();
+    return false;
+}
+
+bool AP_Compass_IIS2MDC::check_whoami()
+{
+    uint8_t whoami = 0;
+    if (!_dev->read_registers(IIS2MDC_ADDR_WHO_AM_I, &whoami, 1)){
+        return false;
+    }
+
+    return whoami == IIS2MDC_WHO_AM_I;
+}
+
+void AP_Compass_IIS2MDC::timer()
+{
+    struct PACKED {
+        uint8_t xout0;
+        uint8_t xout1;
+        uint8_t yout0;
+        uint8_t yout1;
+        uint8_t zout0;
+        uint8_t zout1;
+        uint8_t tout0;
+        uint8_t tout1;
+    } buffer;
+
+    const float range_scale = 100.f / 65535.f; // +/- 50 Gauss, 16bit
+
+    uint8_t status = 0;
+    if (!_dev->read_registers(IIS2MDC_ADDR_STATUS_REG, &status, 1)) {
+        return;
+    }
+
+    if (!(status & IIS2MDC_STATUS_REG_READY)) {
+        return;
+    }
+
+    if (!_dev->read_registers(IIS2MDC_ADDR_OUTX_L_REG, (uint8_t *) &buffer, sizeof(buffer))) {
+        return;
+    }
+
+    int16_t x = int16_t((buffer.xout1 << 8) | buffer.xout0);
+    int16_t y = int16_t((buffer.yout1 << 8) | buffer.yout0);
+    int16_t z = -int16_t((buffer.zout1 << 8) | buffer.zout0);
+
+    Vector3f field = Vector3f{x * range_scale, y * range_scale, z * range_scale };
```

Comment:
```suggestion
    const Vector3f field{x * range_scale, y * range_scale, z * range_scale };
```

Reply:
this can't be const because `accumulate_sample(field, _instance, 20)` takes a field reference and modifies it